Make NG-Scope SDR parameters configurable

bastian-src opened the pull request:
* [x] Add default config parsing test
* [x] Add partial config (missing fields) parsing test
* [x] Add NG-Scope parameters to confy

bastian-src committed:
Add parse tests

* Add default config parsing test
* Add partial config (missing fields) parsing test

bastian-src committed:
Add ng_sdr_config parameter

Provide the option to pass NG-Scope SDR configuration
to set an SDR's serial number and NG-Scope cell parameter n_id.

Limit to three SDRs (a, b, and c).

bastian-src committed:
Implement [ngcontrol] multi-cell handling

Incorporate config-based ng_sdr_config instead of static
default parameters.

bastian-src committed:
Fix clippy

bastian-src commented:
@justus237 I just finished this PR, can you test it from this feature branch?

Be aware to make a backup of your existing default-config and/or add the following parameters to the existing `ngscope` block:

```
ngscope:
...
  ng_sdr_config:
    ng_sdr_a:
      ng_sdr_a_serial: <some SDR serial number>
      ng_sdr_a_n_id: -1
    ng_sdr_b:
      ng_sdr_b_serial: <some SDR serial number>
      ng_sdr_c_n_id: -1
    ng_sdr_c:
      ng_sdr_c_serial: <some SDR serial number>
      ng_sdr_c_n_id: -1
```

or just a single SDR:

```
ngscope:
...
  ng_sdr_config:
    ng_sdr_a:
      ng_sdr_a_serial: <some SDR serial number>
      ng_sdr_a_n_id: -1
    ng_sdr_b: null
    ng_sdr_c: null
```

If you approve, I'll merge it to `main` :penguin:

justus237 commented:
So the single SDR setup still works, but it does not actually make use of multiple SDRs. Not sure whether this is a limitation of DevicePublisher? At the start UECellTracker only prints a single cell. The `/api/v1/celldata/connected/all` returns a single cell, but network signal guru definitely says that the phone is using CA. Changing the path to `/api/v1/celldata/all` doesn't really help, because then UECellTracker just keeps initializing ng-scope over and over, because all connected and available cells changes quite frequently.

You need to set ng-scope's nof_rf_dev to the number of SDRs and ng-scope throws an error if you try to listen to the same base station (presumably same EARFCN) with two SDRs. As a result, I think it makes sense to assume 2xCA and just maybe modify DevicePublisher? Would need some hints on how to get the App onto the phone though.

justus237 committed:
Adapted CLI args fix by justus237

fix cli args causing null values in config

bastian-src commented:
@justus237 I've added setting  `nof_rf_dev` according to the current number of SDRs (considering the number of cells) as discussed.
Moreover, I've cherry-picked your changes from https://github.com/justus237/UECellTracker/tree/fix/cli-args and applied them to this PR. I can see if there are too many changes now for an actual review, but maybe you can give it another try on the hardware? Also, I'm not "overwriting" the .yaml config anymore.

I've also added tests which verify that parameters, which are not part of the .yaml config, will now be filled with the constat defaults. :whale:

justus237 commented:
For some reason the rnti matcher isn’t running (the pattern also doesnt show up in bmon; I’m not sure I’ll have time to root cause this next week) but ng-scope is configured correctly and starting correctly (currently just manually waiting until the devicepublisher shows the two CA cells near me.. :D but this should realistically be what we get from the API so this LGTM).

bastian-src commented:
@justus237 can you test with a single SDR too?
